### ISSUE TITLE
outer clipboard vs. primary selection blocking (and depending on what interfaces the outer wayland compositor supports)?

### ISSUE THREAD

**Issue by jonleivent** [talex5/wayland-proxy-virtwl#60]
I am attempting to use wayland-proxy-virtwl to help me sandbox apps with a labwc compositor desktop, and I noticed some unusual behavior that I an not sure is a bug.

Is the intention of wayland-proxy-virtwl is to prevent data in the form of the clipboard and primary selection presented to the outer socket from being accessible to clients on the socket wayland-proxy-virtwl manages, but allow clients on the same wayland-proxy-virtwl provided socket to share clipboard and primary selections (and allow both to be accessible outside)?  This is the behavior I see for the clipboard but not the primary selection - the outside primary is accessible inside.  Also, as I'm using a special version of labwc (https://github.com/labwc/labwc/discussions/1002#discussioncomment-6564827), if I block some parts of the protocol from wayland-proxy-virtwl, it surprisingly defeats its ability to block the outside clipboard selection.

Is it the intention of wayland-proxy-virtwl to block the clipboard and primary, or just the clipboard?  Is there a way to block the primary as well?

**Comment by talex5** [comment]
No, the proxy isn't trying to prevent clipboard or selection access. If one of them isn't working, it's just a bug, not a security feature :-)

The proxy is expected to run inside an untrusted VM, so relying on it to prevent apps in the VM from seeing host application data isn't a good idea. That needs to happen outside the VM - ideally in the host compositor, although you could use another (modified) proxy for that. But at the moment there's no support for that.

**Comment by jonleivent** [comment]
Yes, I was instead looking for the ability to run this proxy outside a namespace sandbox (bwrap), and bind only the new socket into the sandbox, and have the ability to block wl_data_offers in one direction (those coming from the backing compositor) while still allowing wl_data_offers in the other.

Perhaps I can hack wayland-proxy-virtwl to do this, although my OCaml knowledge was never very detailed.

**Comment by jonleivent** [comment]
It seems like I can easily prevent data_offers for the proxy's clients with just 3 line deletions in relay.ml: the `H.<offering protocol>.receive h ~mime_type ~fd` lines.  Do you see a problem with that?

If this works, it will do nearly what I want - it won't allow data offers to pass among clients of the proxy, however.  But for the most part, each client runs in its own sandbox, so sharing data among clients in the same sandbox is not as important.

**Comment by jonleivent** [comment]
> I've succeeded in adding named clipboards to wayland-proxy-virtwl....

https://github.com/jonleivent/wayland-proxy-clipname
